### PR TITLE
Support "Rules" in the YAMLStoryWriter

### PR DESCRIPTION
**Pull request by degiz**
**Proposed changes**:
- Processing `RuleStep` separately from `StoryStep`

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/shared/core/training_data/structures.py [hunk 1]
```diff
@@ -86,15 +86,13 @@ def __init__(
...
         source_name: Optional[Text] = None,
-        is_rule: bool = None,
     ) -> None:
 
         self.end_checkpoints = end_checkpoints if end_checkpoints else []
         self.start_checkpoints = start_checkpoints if start_checkpoints else []
         self.events = events if events else []
         self.block_name = block_name
         self.source_name = source_name
-        self.is_rule = is_rule
```
Comment (federicotdn) [inline rasa/shared/core/training_data/structures.py:97]: Do you think we could leave this variable as-is? It's useful for checking if we're dealing with rules or not, without having to use `isinstance`.

File: rasa/shared/core/training_data/structures.py [hunk 2]
```diff
@@ -221,23 +218,85 @@ def explicit_events(
 
         return events
 
+    def get_rules_condition(self) -> Optional[List[Event]]:
+        pass
```
Comment (federicotdn) [inline rasa/shared/core/training_data/structures.py:222]:
```suggestion
        return None
```

File: rasa/shared/core/training_data/structures.py [hunk 3]
```diff
@@ -221,23 +218,85 @@ def explicit_events(
...
+
+    def get_rules_steps(self) -> Optional[List[Event]]:
+        pass
```
Comment (federicotdn) [inline rasa/shared/core/training_data/structures.py:225]:
```suggestion
        return None
```

File: rasa/shared/core/training_data/structures.py [hunk 4]
```diff
@@ -221,23 +218,85 @@ def explicit_events(
...
+            "end_checkpoints={!r}, "
+            "events={!r})".format(
+                self.block_name,
+                self.start_checkpoints,
+                self.end_checkpoints,
+                self.events,
+            )
+        )
+
+    def get_rules_condition(self) -> List[Event]:
```
Comment (federicotdn) [inline rasa/shared/core/training_data/structures.py:272]: The return type should match the parent's (`Optional`).

File: rasa/shared/core/training_data/structures.py [hunk 5]
```diff
@@ -221,23 +218,85 @@ def explicit_events(
...
+        """Returns a list of events forming a condition of the Rule. """
+
+        return [
+            event
+            for event_id, event in enumerate(self.events)
+            if event_id in self.condition_events_indices
+        ]
+
+    def get_rules_events(self) -> List[Event]:
```
Comment (federicotdn) [inline rasa/shared/core/training_data/structures.py:281]: Same here!

File: rasa/shared/core/training_data/structures.py [hunk 6]
```diff
@@ -221,23 +218,85 @@ def explicit_events(
...
+        """Returns a list of events forming the Rule, that are not conditions. """
+
+        return [
+            event
+            for event_id, event in enumerate(self.events)
+            if event_id not in self.condition_events_indices
+        ]
+
+    def add_event_as_condition(self, event: Event) -> None:
+        """Adds event to the Rule as part of it's condition.
```
Comment (federicotdn) [inline rasa/shared/core/training_data/structures.py:291]:
```suggestion
        """Adds event to the Rule as part of its condition.
```

File: rasa/shared/core/training_data/story_reader/story_step_builder.py [hunk 7]
```diff
@@ -101,10 +102,17 @@ def add_user_messages(
                     updated_steps.append(copied)
             self.current_steps = updated_steps
 
-    def add_event(self, event) -> None:
+    def add_event_as_condition(self, event) -> None:
```
Comment (federicotdn) [inline rasa/shared/core/training_data/story_reader/story_step_builder.py:105]: Let's add type for `event`.

File: rasa/shared/core/training_data/story_reader/story_step_builder.py [hunk 8]
```diff
@@ -101,10 +102,17 @@ def add_user_messages(
...
+        self.add_event(event, True)
+
+    def add_event(self, event, is_condition=False) -> None:
```
Comment (federicotdn) [inline rasa/shared/core/training_data/story_reader/story_step_builder.py:108]: Same here, and `is_condition`.

File: rasa/shared/core/training_data/story_reader/yaml_story_reader.py [hunk 9]
```diff
@@ -81,17 +81,29 @@ def read_from_file(self, filename: Union[Text, Path]) -> List[StoryStep]:
...
+    def read_from_string(self, string: Text) -> List[StoryStep]:
+        """Read stories or rules from a string.
+
+        Args:
+            string: Unprocessed YAML file content.
+
+        Returns:
+            `StoryStep`s read from `string`.
+        """
+
```
Comment (federicotdn) [inline rasa/shared/core/training_data/story_reader/yaml_story_reader.py:99]:
```suggestion
```

File: rasa/shared/core/training_data/story_writer/yaml_story_writer.py [hunk 10]
```diff
@@ -89,25 +111,28 @@ def process_story_step(self, story_step: StoryStep) -> OrderedDict:
...
+            return self.process_or_utterances(event)
+        if isinstance(event, UserUttered):
+            return self.process_user_utterance(event)
+        if isinstance(event, ActionExecuted):
+            return self.process_action(event)
+        if isinstance(event, SlotSet):
+            return self.process_slot(event)
+        if isinstance(event, ActiveLoop):
+            return self.process_active_loop(event)
+
```
Comment (federicotdn) [inline rasa/shared/core/training_data/story_writer/yaml_story_writer.py:135]:
```suggestion
        if isinstance(event, ActiveLoop):
            return self.process_active_loop(event)
        return None

```

**Review by federicotdn — APPROVED** [review]
Tested it locally and it's working nicely! Added some small comments

File: rasa/shared/core/training_data/structures.py (hunk 1 above)
Comment (degiz) [reply rasa/shared/core/training_data/structures.py:97]: I wanted to separate the `Rules` from non-`Rules` completely, just to clearly abstract all rule-related stuff.
The `get_rules_condition` and `get_rules_steps` inside the `StoryStep` class are my mistakes, I'll remove them 😬

File: rasa/shared/core/training_data/structures.py (hunk 2 above)
Comment (degiz) [reply rasa/shared/core/training_data/structures.py:222]: removed.

File: rasa/shared/core/training_data/structures.py (hunk 4 above)
Comment (degiz) [reply rasa/shared/core/training_data/structures.py:272]: removed.

File: rasa/shared/core/training_data/structures.py (hunk 3 above)
Comment (degiz) [reply rasa/shared/core/training_data/structures.py:225]: removed.

**Pull request merged**